### PR TITLE
Data/code sharing and data limitations

### PR DESCRIPTION
**Pull request by agitter**
I'm trying to finish a complete draft of the Discussion section.  These sub-sections could definitely use feedback.  I confirmed we'll be getting a draft of the evaluation sub-section soon so I didn't write that.  I also removed other Discussion prompts.

@brettbj I'd like to briefly touch on the diet networks paper (#140) here.  Can you please help me describe this in a sentence or two?  What I have is not very clear.

@cgreene I'm closing #172 and we can include any relevant papers in this sub-section.

If there are build errors, I'll clean them up later today.

### CODE REVIEW

File: sections/06_discussion.md [hunk 1]
```diff
@@ -183,32 +187,83 @@ multitudinous layers and architectures available
...
+network can be quickly fine-tuned on new data and used in transfer learning,
+both discussed below.  `TODO: bring up image-encoding for variant classification
+that used pre-trained image classifier?` "Model zoos", collections of
+pre-trained models, are not yet common in biomedical domains but have started to
+appear in genomics applications [@tag:Angermueller2016_single_methyl
+@tag:Dragonn]. `TODO: other examples, possibly from Categorize or Treat?`
+
+DeepChem [@tag:AltaeTran2016_one_shot @tag:DeepChem @tag:Wu2017_molecule_net]
+and DragoNN [@tag:Dragonn] exemplify the benefits of sharing code under an open
+source license and pre-trained models. `TODO: it would be great to add more
```
Comment (cgreene) [inline sections/06_discussion.md]: Source code:
https://github.com/greenelab/deep-review/issues/213 - Tagging @spoilt333 - don't think the raw data is here (looks like `/data/` path but not 100% sure). - did not see a license
https://github.com/greenelab/deep-review/issues/266 - tagging @zhangyan32 - looks like source but not models + data - did not see a license
https://github.com/greenelab/deep-review/issues/217 - source code, data can be downloaded - tagging @uci-cbcl - has license (GPL)
https://github.com/greenelab/deep-review/issues/171 tagging @CampagneLaboratory - this one has some docs, but not data/models if i'm seeing things correctly - has license (Apache)
https://github.com/greenelab/deep-review/issues/155 - code yes w/ docs. Don't think data. Tagging @traversc - just a quick note as I look through the source code I think some .pyc files made it in. not sure if that's intended. - no license
https://github.com/greenelab/deep-review/issues/150 - contains pre-trained models - tagging @raghavchalapathy - did not see a license
https://github.com/greenelab/deep-review/issues/32 - also has source code from @cui-cbcl + example file + data download + very nice readme - Custom license
https://github.com/greenelab/deep-review/issues/24 - source code + data [not version controlled]

That last one highlights to me that we don't _mention_ version control here but we probably should recommend that explicitly.

**Comment by agitter** [comment]
@cgreene thanks for all the helpful suggestions.  I should be able to work on these Monday.

File: sections/06_discussion.md (hunk 1 above)
Comment (zhangyan32) [inline sections/06_discussion.md]: My repo is source code, and we are still working on it now.

**Comment by brettbj** [comment]
@agitter Will do tonight

**Comment by brettbj** [comment]
@agitter @cgreene - needed a full reread last night, will add the couple of sentences after a few meetings this morning

**Comment by agitter** [comment]
@brettbj no rush, I won't get to this until today or tomorrow most likely because I shifted my efforts to drafting #370

File: sections/06_discussion.md [hunk 2]
```diff
@@ -183,32 +188,83 @@ multitudinous layers and architectures available
...
+domains, some of the best training data has been generated privately, for
+example, high-throughput chemical screening data at pharmaceutical companies.
+There is little expectation or incentive for this private data to be shared.
+`TODO: this last line is fairly pessimistic, is there something better to say?
+Is it possible to create the right incentives and/or protect proprietary data?`
+
+Code sharing and open source licensing is essential for continued progress in
+this domain.  Even though some journals have improved requirements about code
+availability at the time of publication, there remain many opportunities to do
+better.  Preprints are firmly established in many areas of machine learning and
```
Comment (agapow) [inline sections/06_discussion.md]: Hmmm. I feel that preprints - great as they are - is perhaps getting a little out of scope and are qualitatively different to code and data sharing. Just trying to pick out the deep learning specific material here.

Comment (agitter) [inline sections/06_discussion.md]: Perhaps it is out of scope.  I brought it up because it has been a point of personal frustration.  More than once, there has been an exciting preprint on some biomedical deep learning topic that my group wanted to test out or build upon.  If the code isn't available, we're stuck.  Sometimes we don't even want to run the code, we just want to understand minute details of the training that aren't written down in the paper.  Without code, we're left guessing.

Part of what I wanted in this section is a call to go beyond the accepted practices in the community.  That includes sharing code when not required and highlighting example repositories that make the extra effort to provide pre-trained models, data pre-processing scripts, tutorials, etc.

File: sections/06_discussion.md [hunk 3]
```diff
@@ -183,32 +188,83 @@ multitudinous layers and architectures available
...
+data that would help to power deep learning in the domain. Efforts are underway
+to recognize those who promote an ecosystem of rigorous sharing and analysis
+[@doi:10.1038/ng.3830].
+
+The sharing of high-quality, labeled datasets will be especially valuable.  In
+addition, researchers who invest time to preprocess datasets to be suitable for
+deep learning can make the preprocessing code (e.g., Basset
+[@tag:Kelley2016_basset]) and cleaned data (e.g., MoleculeNet
+[@tag:Wu2017_molecule_net]) publicly available to catalyze further research. As
+discussed in the Categorize section, there are complex privacy and legal issues
```
Comment (enricoferrero) [inline sections/06_discussion.md]: Do we actually refer to that section as "the Categorize section" in the paper?

File: sections/06_discussion.md (hunk 2 above)
Comment (agapow) [inline sections/06_discussion.md]: I get it. My feeling is that is a broader issue and might diffuse any DL-specific message. I'll leave this to you. Maybe see how it fits when the paper is more complete and polished.

Comment (enricoferrero) [inline sections/06_discussion.md]: While it might be slightly out of scope, I think we should keep it. It's an important point and it flows well with the discussion. A couple of additional thoughts:
- maybe say explicitly that this recommendation is a general one, not specific to deep learning;
- we could go as far as recommend code sharing platforms such as GitHub (maybe even Zenodo?)

File: sections/06_discussion.md (hunk 3 above)
Comment (agitter) [inline sections/06_discussion.md]: Probably not 😄 

I'm used to that jargon from our issues here and should fix it.

File: sections/06_discussion.md (hunk 2 above)
Comment (agitter) [inline sections/06_discussion.md]: @enricoferrero I support this.  I think to avoid diffusing the DL-specific message further, we should briefly mention these things but then refer to an existing paper.  There is probably a suitable PLoS Comp Bio 10 simple rules or similar article I can cite.

@agapow if it seems out of place when we polish the full Discussion section, I'm willing to drop it.  I prefer to keep it in the initial draft.

File: sections/06_discussion.md [hunk 4]
```diff
@@ -68,46 +72,46 @@ interpretable healthcare features from trained deep models.
...
+There is a counter-example from drug discovery to include as well
+that is related to #55`
+
+Multimodal, multi-task, and transfer learning, discussed in detail below, can
+also combat data limitations to some degree.  There are also emerging
+network architectures, such as Diet Networks for high-dimensional SNP data
+[@tag:Romero2017_diet], that drastically reduce the number of free parameters
+by predicting parameters (weights) as a function of each SNP's embedding
+in some feature space instead of directly optimizing a weight for each SNP.
+`TODO: not very familiar with this paper, can someone describe it better?`
```
Comment (brettbj) [inline sections/06_discussion.md]: How's this (tried to still keep it brief without getting heavily into embeddings) - 

Multimodal, multi-task, and transfer learning, discussed in detail below, can also combat data limitations to some degree. There are also emerging network architectures, such as Diet Networks for high-dimensional SNP data [@tag:Romero2017_diet]. Diet Networks use multiple networks to drastically reduce the number of free parameters by first flipping the problem and tasking a network to predict parameters (weights) for each input (SNP) to learn a feature embedding. This embedding (i.e. PCA, per class histograms, x2vec style) can be learned directly from the input data, but has the advantage that it can also be learned from other datasets or from domain knowledge. Additionally, in this task, the features are the examples, an important fact when it is typical to have 500k+ SNPs and only a few thousand patients. Finally, this embedding is of a much lower dimension, allowing for a large reduction in the number of free parameters. In the example given, the authors reduced the number of free parameters from 30M to 50K, a factor of 600.

File: sections/06_discussion.md [hunk 5]
```diff
@@ -68,46 +72,46 @@ interpretable healthcare features from trained deep models.
...
-*Bengio diet networks paper*
-
+A lack of large-scale, high-quality, labeled training data has impacted deep
+learning in nearly all applications we have discussed, from healthcare to
+genomics to drug discovery.  The challenges of training complex, high-parameter
+neural networks from few examples are obvious, but uncertainty in the labels of
+training instances can be just as problematic.  In genomics, for example,
+labeled data may be derived from an experimental assay that has known and
+unknown technical artifacts, biases, and error profiles.  It is possible to
+weight training examples or construct Bayesian models to account for uncertainty
```
Comment (agitter) [inline sections/06_discussion.md]: I made #377 for you to add these lines.  You'll do a better job of getting the example correct.

This is along the lines of what I meant by "weight training examples", but I didn't explain that phrase well.

File: sections/06_discussion.md [hunk 6]
```diff
@@ -68,46 +72,46 @@ interpretable healthcare features from trained deep models.
...
+in the data, but this is rarely done in practice. `TODO: Is this even true?
+Would like feedback here.`
+
+For some types of data, especially images, it is straightforward to augment
+training datasets by splitting one labeled example into multiple examples. An
+image can easily be rotated or inverted and retain its label `TODO: does someone
+have examples to cite here? Issue #163?`.  3D MRI and 4D fMRI (with time as a
```
Comment (agitter) [inline sections/06_discussion.md]: Added #163

File: sections/06_discussion.md [hunk 7]
```diff
@@ -68,46 +72,46 @@ interpretable healthcare features from trained deep models.
...
+dimension) data can be decomposed into sets of 2D images [@doi:10.1101/070441].
+This greatly expands the number of training examples in domains where only
+hundreds of subjects are available but artificially treats images from the same
+volume as independent instances.  CellCnn trains a model to recognize rare cell
+populations in single-cell data by creating training instances that consist of
+random subsets of cells that are randomly sampled with replacement from the
+full dataset [@tag:Arvaniti2016_rare_subsets].
+
+Simulated or semi-synthetic training data has also been employed in multiple
+biomedical domains. `TODO:  simulated data: #5 #99 #293, maybe #117 and #197.
```
Comment (agitter) [inline sections/06_discussion.md]: I opened #378 for that paper.  I'll see if I need it.

File: sections/06_discussion.md (hunk 4 above)
Comment (agitter) [inline sections/06_discussion.md]: @brettbj thanks, I added this.  I want to rephrase the `x2vec style` or remove the example embedding types and will use everything else as-is.

File: sections/06_discussion.md [hunk 8]
```diff
@@ -183,32 +187,83 @@ multitudinous layers and architectures available
...
+addition, researchers who invest time to preprocess datasets to be suitable for
+deep learning can make the preprocessing code (e.g., Basset
+[@tag:Kelley2016_basset]) and cleaned data (e.g., MoleculeNet
+[@tag:Wu2017_molecule_net]) publicly available to catalyze further research. As
+discussed in the Categorize section, there are complex privacy and legal issues
+involved in sharing patient data and deep learning models trained with it but
+also recent techniques to mitigate these concerns.  Furthermore, in some
+domains, some of the best training data has been generated privately, for
+example, high-throughput chemical screening data at pharmaceutical companies.
+There is little expectation or incentive for this private data to be shared.
```
Comment (agitter) [inline sections/06_discussion.md]: I used this.  There is an example where Merck provided data for a Kaggle competition, but I don't want to get into too much detail.

File: sections/06_discussion.md [hunk 9]
```diff
@@ -183,32 +187,83 @@ multitudinous layers and architectures available
...
+`TODO: this last line is fairly pessimistic, is there something better to say?
+Is it possible to create the right incentives and/or protect proprietary data?`
+
+Code sharing and open source licensing is essential for continued progress in
+this domain.  Even though some journals have improved requirements about code
+availability at the time of publication, there remain many opportunities to do
```
Comment (agitter) [inline sections/06_discussion.md]: I don't think I want to focus on journal requirements, so I'll rephrase to state that we encourage best practices for software regardless of what is or isn't required by publishers.

File: sections/06_discussion.md (hunk 2 above)
Comment (agitter) [inline sections/06_discussion.md]: @agapow I revised this paragraph to try to keep it more focused on the core element of code sharing.  I also cite http://science.sciencemag.org/content/354/6317/1240.full to expand on best practices so we don't get too off topic here.

@cgreene and @enricoferrero, please see if you still like the overall message (once I push).

File: sections/06_discussion.md [hunk 10]
```diff
@@ -183,32 +187,83 @@ multitudinous layers and architectures available
...
+Because many deep learning models are often built using one of several popular
+deep learning frameworks, it is also possible to directly share trained
+predictive models.  The availability of pre-trained models can accelerate
+research, with image classifiers as an apt example.  A pre-trained neural
+network can be quickly fine-tuned on new data and used in transfer learning,
+both discussed below.  `TODO: bring up image-encoding for variant classification
+that used pre-trained image classifier?` "Model zoos", collections of
+pre-trained models, are not yet common in biomedical domains but have started to
+appear in genomics applications [@tag:Angermueller2016_single_methyl
+@tag:Dragonn]. `TODO: other examples, possibly from Categorize or Treat?`
```
Comment (agitter) [inline sections/06_discussion.md]: @c two paragraphs above I have "As discussed in the Categorize section, there are complex privacy and legal issues involved in sharing patient data and deep learning models trained with it but also recent techniques to mitigate these concerns."  Do you suggest I move that to this paragraph?

File: sections/06_discussion.md (hunk 1 above)
Comment (agitter) [inline sections/06_discussion.md]: #171 is a good example with tutorials, license, etc.  I think #24 does have a GitHub repo https://github.com/uci-cbcl/D-GEX

File: sections/06_discussion.md (hunk 10 above)
Comment (cgreene) [inline sections/06_discussion.md]: What about changing above to:
"There are complex privacy and legal issues involved in sharing patient data that cannot be ignored."

Then here: "Deep learning models can also be attacked to identify examples used in training. We discuss this issue as well as recent techniques to mitigate these concerns... [this refers to categorize]"

I just want to make sure that nobody drops in here, reads that they should share their model, and then shares a model that leads to inadvertent release of something they thought was private.

**Comment by agitter** [comment]
I believe I addressed all of the helpful comments above.  Let me know if I missed something.

@cgreene we can do a final review and merge now if everything looks okay.  I have one TODO remaining regarding the simulated data paragraph, but I can make a separate small pull request for that.

**Comment by cgreene** [comment]
This LGTM :+1:. The outstanding issues appear to be noted, and we can go through with more fine grained revisions after a merge if necessary.

File: sections/06_discussion.md (hunk 10 above)
Comment (agitter) [inline sections/06_discussion.md]: I'll make this edit, then merge
